lib/palettes_base.hpp

```diff
@@ -0,0 +1,227 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-443271.
+//
+// This file is part of the GLVis visualization tool and library. For more
+// information and source code availability see https://glvis.org.
+//
+// GLVis is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef GLVIS_PALETTESBASE_HPP
+#define GLVIS_PALETTESBASE_HPP
+
+#include "gl/types.hpp"
+
+#include <array>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <memory>
+#include <vector>
+#include <iomanip>
+
+using namespace std;
+
+struct RGBAf
+{
+   float r, g, b, a;
+
+   constexpr RGBAf(float r = 0.0, float g = 0.0, float b = 0.0, float a = 1.0)
+      : r(r), g(g), b(b), a(a) {}
+
+   void Print(ostream& os = cout) const;
+
+   array<float, 4> AsArray() const { return {r, g, b, a}; }
+};
+
+
+class Palette
+{
+public:
+   const string name;
+
+   /// Constructors
+   Palette(const string& name) : name(name) {}
+
+   /// Constructor from vector of RGBAf
+   Palette(const string& name, const vector<RGBAf>& colors)
+      : name(name), colors(colors) {}
+
+   /// Constructor from Nx3 array
+   template <size_t N>
+   Palette(const string& name, const array<array<float,3>,N>& arr);
+
+   /// Constructor from Nx4 array
+   template <size_t N>
+   Palette(const string& name, const array<array<float,4>,N>& arr);
+
+   /// Get size
+   int Size() const { return colors.size(); }
+
+   /// Add color to palette
+   void AddColor(float r, float g, float b, float a = 1.0);
+
+   /// Print each color of this palette to a stream
+   void Print(ostream& os = cout) const;
+
+   /// Get color at index i (optionally, use reversed order)
+   RGBAf Color(int i, bool reversed = false) const;
+
+   /// Get all colors as a vector of float arrays
+   vector<array<float,4>> GetData(bool reversed = false) const;
+
+   /// Are any alpha != 1.0?
+   bool IsTranslucent() const;
+
+private:
+   vector<RGBAf> colors;
+};
+
+using TexHandle = gl3::resource::TextureHandle;
+/// Generates the texture data for a given palette, to be used in OpenGL
+class Texture
+{
+public:
+   // What type of texture is this (discrete, smooth, alphamap)
+   enum class TextureType
+   {
+      DISCRETE = 0,
+      SMOOTH = 1,
+      ALPHAMAP = 2,
+   };
+
+   /// Empty constructor
+   Texture() {}
+
+   /// Constructor - generates texture
+   Texture(const Palette* palette,
+           TextureType textype = TextureType::DISCRETE,
+           int cycles = 1, int colors = 0);
+
+   /// Constructor for alphamap
+   Texture(float matAlpha, float matAlphaCenter);
+
+   /// Initialize Static GL parameters
+   void InitStaticGL();
+
+   /// Get texture size.
+   int Size() { return tsize; }
+
+   /// Get the GL texture
+   GLuint Get() const { return texture; }
```

Comment:
The naming is a bit inconsistent here, either I would use `GetGLTexture()` or rename the method below to only `Generate()`.